csrc/device_lower/analysis/thread_predicate.cpp

```diff
@@ -494,6 +502,179 @@ class RedundantUseAnalysis : BackwardVisitor {
+              std::distance(root_domain_.begin(), outer_iter));
+        }
+        if (inner_iter != root_domain_.end()) {
+          merged_root_domains.emplace_back(*inner_iter);
+          index_root_domain.emplace_back(
+              std::distance(root_domain_.begin(), inner_iter));
+        }
+      } else {
+        // current analysis of predication is only valid if all the exprs
+        // between this lead domain and root domains are merge
+        return std::vector<IterDomain*>();
+      }
+    }
+    // The following sort is added because in NVFuserTest.FusionIssue2076_CUDA
+    // the order is [I3, I1, B2] while the correct order should be [I1, B2, I3]
+    size_t n_elements = merged_root_domains.size();
+    TORCH_INTERNAL_ASSERT(
+        n_elements, "The number of merged root domains should > 0");
+    std::vector<int> indices(n_elements);
+    std::iota(indices.begin(), indices.end(), 0);
+    std::sort(indices.begin(), indices.end(), [&](int a, int b) {
+      return index_root_domain.at(a) < index_root_domain.at(b);
+    });
+    std::vector<IterDomain*> merged_root_domains_sorted(n_elements);
+    for (size_t i = 0; i < n_elements; ++i) {
+      merged_root_domains_sorted.at(i) = merged_root_domains.at(indices.at(i));
+    }
+
+    return merged_root_domains_sorted;
+  }
+
+  // Get the index of the leaf domain if we skip the broadcasted root domains
+  Val* getIndexWithoutBroadcast(
+      const std::vector<IterDomain*>& merged_root_domains,
+      ParallelType pt) {
+    const int ndim = (int)merged_root_domains.size();
+    // get the stride if we index the leaf domain using its root domains
+    std::vector<Val*> root_stride(ndim);
+    root_stride.at(ndim - 1) = GpuLower::current()->kernel()->oneVal();
+    for (int i = ndim - 2; i >= 0; i--) {
+      auto pre_crd = merged_root_domains.at(i + 1);
+      Val* pre_extent = pre_crd->isBroadcast()
+          ? concretized_broadcast_root_domains_.at(pre_crd)->extent()
+          : pre_crd->extent();
+      root_stride.at(i) = IrBuilder::mulExpr(root_stride.at(i + 1), pre_extent);
+    }
+    // convert the linear index of the leaf domain to the indices of the root
+    // domains
+    Val* remaining_index = NamedScalar::getParallelIndex(pt);
+    std::vector<Val*> root_indices(ndim);
+    for (int i = 0; i < ndim; i++) {
+      root_indices.at(i) =
+          IrBuilder::divExpr(remaining_index, root_stride.at(i));
+      remaining_index = IrBuilder::modExpr(remaining_index, root_stride.at(i));
+    }
+
+    // get the index of the leaf domain if we skip the broadcasted root domains
+    Val* index_without_broadcast = IrBuilder::create<Int>(0);
+    for (int i = 0; i < ndim; i++) {
+      if (!merged_root_domains.at(i)->isBroadcast()) {
+        index_without_broadcast = IrBuilder::addExpr(
+            index_without_broadcast,
+            IrBuilder::mulExpr(root_indices.at(i), root_stride.at(i)));
+      }
+    }
+    return index_without_broadcast;
```

Comment:
Instead of generating the index that corresponds to the leaf ID, it seems it's more intuitive to just use the index of the broadcast root ID and create a predicate as `index_of_the_broadcast_root_id == 0`

Reply:
Sounds like if we have multiple merged broadcast root IDs, then the condition will be `index_of_the_broadcast_root_id0 == 0 && index_of_the_broadcast_root_id1 == 0 && ...`. Is this prefered?

Reply:
The logic of generating the predicate would be more straightforward. The generated predicate may end up having multiple conditions, and it appears to me easier to reason about.

Reply:
sure. I'll change to that.